### ISSUE TITLE
Viewports + controller input

### ISSUE THREAD

**Issue by shawna-p** [renpy/renpy#5557]
I've been investigating ways to make games more accessible using the keyboard and controllers, and one kind of displayable that's difficult to use with this control scheme is viewports.

In many games which use controllers, if a viewport has a list of items, advancing through the items automatically advances the scroll of the viewport to keep the focused item visible. In particular there are two main approaches to this - one where the viewport is "paged" such that reaching the end of the visible items and advancing brings the next item to the top of the viewport (like paged scrolling), and the second more common version where focusing the next item causes the viewport to scroll just enough to make that item fully within the viewport area (and often doesn't scroll at all when switching focus between items which are already fully visible). So, if you're scrolling down a viewport with a vbox, hitting "down" would ensure that at least the bottom edge of the focused item is at the bottom edge of the viewport so it's visible. This is sometimes instantaneous and sometimes accompanied by a short animation to drift from one scroll value to the next.

Currently this is not possible to do with the focus system, which only knows about the items in the viewport which are visible and doesn't consider scrolling the viewport in any way to focus items currently not rendered (and thus will start focusing buttons outside of the viewport if possible).

**Comment by renpytom** [comment]
This is a hard one to fix, hard enough I'm not sure I know a great way to do it.

**Comment by shawna-p** [comment]
I tried an approach with scrolling and the focus system which ultimately wasn't suitable; the main issue is not having position information of the contents in the viewport relative to the viewport itself.
I'm working on a second approach which is restricted to hbox, vbox, and grid, and uses the saved offset information to calculate a child to focus. One thing that would help with that approach is if the boxes/grid saved the children sizes to a field alongside the offsets so I can access those after it's rendered. 
It's still probably best suited as a plugin at this point, but I think most scrollable viewports use boxes/grid for arrangement so it should solve part of the desired use cases.

**Comment by renpytom** [comment]
I'm having some ideas, but thios won't be done before 8.3 at the earliest. Potentially later if they don't pan out. (Pardon the pun.)

**Comment by shawna-p** [comment]
Linking #2224 #5363 and #3979 as all have similar-seeming focus issues with controllers.

**Comment by feltcute** [comment]
Voicing support of this feature,

I've been wanting this for these last few weeks in a viewport heavy project. As a prelude toward generally better non-mouse/touch support, I found my mouse and touch input needs simply differed from keyboard and controller various circumstances, so I turned to a persistent variable for a displayable constantly checking the last used input to change the UI as needed.

`            if ev.type == pygame.MOUSEMOTION:
                persistent.lastInput = "Mouse"`
                
And so forth. It was looking bright up till exploring this focus based scrolling behavior, it seemed unapproachable and beyond me. I was rather close to just going for a dynamically swapped interface with page buttons akin to the out of the box save menu, but if this viewport feature gains ground as either a plugin or native feature in a future release, I'd definitely prefer that.

**Comment by egandro** [comment]
I think this is related to that feature.

When using a controller e.g. PS4 or Xbox, I would love to see a feature to make an item in a menu having the default focus.

At the moment you have to do at least one d-pad movement to highlight an element.

**Comment by sunrisesarsaparilla** [comment]
If this does get implemented, I have a UI suggestion: instead of advancing the menu items when the user would surpass the boundary, it's more user-friendly to advance the items one item menu before surpassing the boundary, because this informs the user that there are still more items below.

For example, if we have a menu like this:

```
------------ viewport boundary
menu item A
menu item B
menu item C
menu item D <----selected by user
menu item E
------------ viewport boundary
menu item F (which the user can't see)
```

If the user navigates down, we see:

```
menu item A (which the user can't see)
------------ viewport boundary
menu item B
menu item C
menu item D
menu item E <----selected by user
menu item F
------------ viewport boundary
```

If the user navigates down again, we see:

```
menu item A (which the user can't see)
------------ viewport boundary
menu item B
menu item C
menu item D
menu item E
menu item F <----selected by user
------------ viewport boundary
```

With this design, the bottom item of a list can only be selected is when there are no more items below it, which serves as an additional signal to the user that they've reached the boundary.

Danganronpa does this: https://github.com/user-attachments/assets/c26f1b58-4115-4858-8748-5207525e9ad3

**Comment by shawna-p** [comment]
I agree with @sunrisesarsaparilla, though I'll rephrase it that you should be able to provide a scroll "bonus offset" (potentially a different one for each direction) rather than tying it to "seeing the next button" - I ran into this problem myself in my own experiments where it's essential to be able to scroll up an extra amount if you have, say, a label and then a button e.g.

Music volume:
-------------o--------

and you need to be able to see the Music Volume label when you highlight the bar. A bonus offset would neatly solve both of these problems and be more flexible.

**Comment by jsfehler** [comment]
I was able to get a hacky version of this working. The focus system still doesn't know about displayables hidden by the viewport. This approach requires a consistent id for the displayables in the viewport. If a limit is reached, a scroll is triggered:

- Ensure every button in the viewport has an id with a consistent and increasing number (e.g.: button_0, button_1, button_2)
- Create "Up" and "Down" Actions which store the viewport's id and the number at which scrolling should start (e.g.: If you want to start scrolling after the 7th button is passed, store the number 7)
- Map those Actions to keyboard keys/controller buttons
- When called, `renpy.display.focus.focus_nearest()` is called in the appropriate direction
- `renpy.display.focus.get_focused()` is called to fetch the id of the focused button
- The button's id number is checked. If it's above the limit, scroll down, if it's under, scroll up.
- Scrolling is then handled by a YScrollValue Action
